### PR TITLE
initial commit of token bucket functionality

### PR DESCRIPTION
**Pull request by kujenga**
Implementation of the [token bucket](https://en.wikipedia.org/wiki/Token_bucket) algorithm backed by appengine's memcache for reliable rate limiting across instances.

### CODE REVIEW

File: membucket/membucket.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+		return 0, err
+	}
+	return buk.Tokens, nil
+}
+
+// Take attempts to pass a request through the rate limit
+// and retries if a concurrency conflict occurs
+func (b *Bucket) Take(amt int64) error {
+	const max = 5
+	var counter int64
```
Comment (doykle) [inline membucket/membucket.go]: Should counter be incremented in the loop?

Comment (kujenga) [inline membucket/membucket.go]: yes!

File: membucket/membucket.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+	"golang.org/x/net/context"
+	"google.golang.org/appengine"
+	"google.golang.org/appengine/memcache"
+)
+
+var (
+	// MembucketNamespace is the namespace used for memcache operations
+	MembucketNamespace = "membucket"
+
+	// ErrNonConformant inticates that there are not enough tokens
```
Comment (AndyHaskell) [inline membucket/membucket.go]: Typo here (indicates)

File: membucket/membucket.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+	var buk mb
+	item, err := memcache.JSON.Get(b.ctx, b.Key, &buk)
+	if err != nil && err != memcache.ErrCacheMiss {
+		return err
+	} else if err == memcache.ErrCacheMiss {
+		// no bucket was perviously set, so we set a default item
+		return memcache.JSON.Add(b.ctx, &memcache.Item{
+			Key: b.Key,
+			Object: &mb{
+				// we assume that the amt is less than the maz bucket size. amt will usually be 1
```
Comment (AndyHaskell) [inline membucket/membucket.go]: max

File: membucket/membucket.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+	_, err := memcache.JSON.Get(b.ctx, b.Key, &buk)
+	if err != nil {
+		return 0, err
+	}
+	return buk.Tokens, nil
+}
+
+// Take attempts to remove the specified number of tokens from the bucket.
+// If there is contention on the bucket itself, it will retry up to 15 times
+// and return an nternal memcache error if it is not able to complete the
```
Comment (AndyHaskell) [inline membucket/membucket.go]: internal

File: membucket/membucket.go [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+// internal format for the information stored in memcache
+// fields must be public for json serialization
+type mb struct {
+	Tokens   int64 `json:"tok"`
+	LastFill int64 `json:"lt"` // unix nano time
+}
+
+// New returns a newly created bucket with the specified parameters.
+// It should be called to construct a new Bucket which acts as a "Client"
+// to the status of that bucket. Rate is measurec in tokens/second
```
Comment (AndyHaskell) [inline membucket/membucket.go]: measured

**Comment by doykle** [comment]
This looks great! Can we merge it?

**Comment by kujenga** [comment]
go for it!

**Pull request merged**